components/brave_ads/core/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v3.cc

```diff
@@ -0,0 +1,142 @@
+    return std::move(callback).Run(/*had_opportunity*/ false, {});
+  }
+
+  GetBrowsingHistory(std::move(user_model), ad_events, std::move(callback));
+}
+
+void EligibleAdsV3::GetBrowsingHistory(
+    targeting::UserModelInfo user_model,
+    const AdEventList& ad_events,
+    GetEligibleAdsCallback<CreativeNotificationAdList> callback) {
+  AdsClientHelper::GetInstance()->GetBrowsingHistory(
+      features::GetBrowsingHistoryMaxCount(),
+      features::GetBrowsingHistoryDaysAgo(),
+      base::BindOnce(&EligibleAdsV3::GetEligibleAds, base::Unretained(this),
+                     std::move(user_model), ad_events, std::move(callback)));
+}
+
+void EligibleAdsV3::GetEligibleAds(
+    targeting::UserModelInfo user_model,
+    const AdEventList& ad_events,
+    GetEligibleAdsCallback<CreativeNotificationAdList> callback,
+    const BrowsingHistoryList& browsing_history) {
+  database::table::CreativeNotificationAds database_table;
+  database_table.GetAll(base::BindOnce(
+      &EligibleAdsV3::OnGetEligibleAds, base::Unretained(this), user_model,
+      ad_events, browsing_history, std::move(callback)));
+}
+
+void EligibleAdsV3::OnGetEligibleAds(
+    const targeting::UserModelInfo& user_model,
+    const AdEventList& ad_events,
+    const BrowsingHistoryList& browsing_history,
+    GetEligibleAdsCallback<CreativeNotificationAdList> callback,
+    const bool success,
+    const SegmentList& /*segments*/,
+    const CreativeNotificationAdList& creative_ads) {
+  if (!success) {
+    BLOG(1, "Failed to get ads");
+    return std::move(callback).Run(/*had_opportunity*/ false, {});
+  }
+
+  if (creative_ads.empty()) {
+    BLOG(1, "No eligible ads");
+    return std::move(callback).Run(/*had_opportunity*/ false, {});
+  }
+
+  const CreativeNotificationAdList eligible_creative_ads =
+      FilterCreativeAds(creative_ads, ad_events, browsing_history);
+  if (eligible_creative_ads.empty()) {
+    BLOG(1, "No eligible ads out of " << creative_ads.size() << " ads");
+    return std::move(callback).Run(/*had_opportunity*/ false, {});
+  }
+
+  absl::optional<CreativeNotificationAdInfo> creative_ad;
+  creative_ad = MaybePredictAdUsingEmbeddings<CreativeNotificationAdInfo>(
+      user_model, eligible_creative_ads);
+
+  if (!creative_ad) {
+    BLOG(1, "No eligible ads out of " << creative_ads.size() << " ads");
+    return std::move(callback).Run(/*had_opportunity*/ false, {});
+  }
+
+  BLOG(1, eligible_creative_ads.size()
+              << " eligible ads out of " << creative_ads.size() << " ads");
+
+  std::move(callback).Run(/*had_opportunity*/ false, {*creative_ad});
+}
+
+CreativeNotificationAdList EligibleAdsV3::FilterCreativeAds(
+    const CreativeNotificationAdList& creative_ads,
+    const AdEventList& ad_events,
+    const BrowsingHistoryList& browsing_history) {
+  if (creative_ads.empty()) {
+    return {};
+  }
+
+  CreativeNotificationAdList filtered_creative_ads;
+
+  std::copy_if(creative_ads.cbegin(), creative_ads.cend(),
```

Comment:
can this filtering be done using `ExclusionRules`?

Reply:
Agreed this should be an exclusion rule, good catch

Reply:
for other types of creative ads we just duplicated v2 logic. For notifications we have this extra requirement, although I couldn't add this extra requirement to the current `notification_ad_exclusion_rules` because it only applies to v3, that's why the custom extra filter. Should I create a custom exclusion_rules class for v3?

Reply:
The exclusion rule can check the version param and if not 3 then the exclusion rule should just return false so that the ad is not excluded

Reply:
added exclusion rule with conditioning on notification ad serving version 👍